arch/xtensa/src/common/xtensa_initialstate.c

```diff
@@ -100,6 +124,71 @@ void up_initial_state(struct tcb_s *tcb)
   xcp->regs[REG_A1] = (uint32_t)tcb->stack_base_ptr + /* Physical top of stack frame    */
                                 tcb->adj_stack_size;
 
+  /* Each task access the TLS variables using the THREADPTR register plus an
+   * offset to obtain the address of the variable.
+   *
+   * TLS layout at link-time, where 0xNNN is the offset that the linker
+   * calculates to a particular TLS variable.
+   *
+   * LOW ADDRESS
+   *         |---------------------------|   Linker Symbols
+   *         | Section                   |   --------------
+   *         | .flash.rodata             |
+   *      0x0|---------------------------| <- _rodata_reserved_start
+   *       ^ | Other Data                |
+   *       | |---------------------------| <- _thread_local_start
+   *       | | .tbss                     | ^
+   *       V |                           | |
+   *   0xNNN | int example;              | | tls_area_size
+   *         |                           | |
+   *         | .tdata                    | V
+   *         |---------------------------| <- _thread_local_end
+   *         | Other data                |
+   *         | ...                       |
+   *         |---------------------------|
+   * HIGH ADDRESS
+   *
+   * Calculate the THREADPTR register's initialization value based on the
+   * link-time offset and the TLS area allocated on the stack.
+   *
+   * HIGH ADDRESS
+   *         |---------------------------| <- top of the Stack
+   *         |                           |
+   *         |                           |
+   *         |                           |
+   *         |                           |
+   *         |---------------------------|
+   *         | .tdata (*)                |
+   *       ^ | int example;              |
+   *       | |                           |
+   *       | | .tbss (*)                 |
+   *       | |---------------------------| <- start of TLS area
+   * 0xNNN | |                           | ^
+   *       | |                           | |
+   *       | |           ...             | | (_thread_local_start -
+   *       | |                           | |   _rodata_reserved_start)
+   *       | |                           | |  + align_up(TCB_SIZE,
+   *       | |                           | |             tls_section_align)
+   *       | |                           | |
+   *       | |                           | V
+   *       V |                           | <- threadptr register's value
+   *
+   * LOW ADDRESS
+   *
+   * Note: Xtensa is slightly different compared to the RISC-V port as there
+   * is an implicit aligned TCB_SIZE added to the offset.
+   * (search for 'tpoff' in elf32-xtensa.c in BFD):
+   *  - "offset = address - tls_section_vma +
+   *              align_up(TCB_SIZE, tls_section_align)"
+   *    - TCB_SIZE is hardcoded to 8
+   */
+
+#ifdef CONFIG_SCHED_THREAD_LOCAL
+  xcp->regs[REG_THREADPTR] = sched_tls_data -
```

Comment:
why xtensa need so complex compute, but riscv is very simple?

Reply:
Because of the way the associated register is used. On Xtensa, the `THREADPTR` plus an offset (let's name it `0xNNN`) is used to calculate the location of the TLS data. This offset is calculated at link-time and we need to add marks to consider it when calculating the `THREADPTR` value.

This is explained in this same file:https://github.com/apache/nuttx/pull/15269/files#diff-6ea32ef16de4dfd7e2971ae010a3c4c82b09751ab8acd8bf0e055baba1b93961R127

Any suggestions on how to improve the explanation?

Reply:
I removed the `sched_tls_data` variable and improved the explanation.